juspay/hyperswitch
# feat(payments_v2): add payment_confirm_intent api endpoint

## Type of Change



## Description

This PR introduces the `payment-confirm-intent` endpoint. This is used to confirm the payment intent with payment method data.

This PR also includes changes to the payment attempt for api v2. Few columns have been added and few have been removed. The major changes that have been made to payment attempt are

- The id type is changed to GlobalId
- Amount related details are stored in a struct in the domain model. This is to ensure that all the operations / getters / setters related to amount are available in a single struct.
- Few fields that have been removed as a part of backwards compatibility removal have been deleted in the database.
- The address details have been de - normalized and stored in the payment attempt table. This is because we will not be reusing the address.

This PR contains only the minimum changes that are required to call the connector. 
- There is a bare minimum implementation of routing which is based on the fallback connector config of the business profile.
- There is no surcharge, decision manager, tokenization and other features. These would be refactored and slowly added to the v2 feature set.
- There is no implementation to save the card. These features would eventually be added to v2.

### Additional Changes

         Add new api endpoint to payments. The openapi reference has been generated. 
<img width="1456" alt="Screenshot 2024-10-11 at 11 00 15 AM" src="https://github.com/user-attachments/assets/93fa11c8-34b6-4621-bdc0-56135791602f">





## Motivation and Context



## How did you test it?

- Create the payment intent
- Confirm the payment intent.
```bash
curl --location 'http://localhost:8080/v2/payments/12345_pay_0192b430fb0272b2a4ef5efdbfb06f68/confirm-intent' \
--header 'x-profile-id: pro_an7sjopN2AXVvhEEYvSm' \
--header 'x-client-secret: 12345_pay_0192b430fb0272b2a4ef5efdbfb06f68_secret_0192b430fb0272b2a4ef5f08095c2b1b' \
--header 'Content-Type: application/json' \
--header 'api-key: pk_dev_78d1c87dc958419587490ece50cbfd0c' \
--data '{
    "payment_method_data": {
            "card": {
                "card_number": "4242424242424242",
                "card_exp_month": "01",
                "card_exp_year": "27",
                "card_holder_name": "John Doe",
                "card_cvc": "100"
            }
    },
    "payment_method_type": "card",
    "payment_method_subtype": "credit"
}'
```

## Checklist
